dev/drupal#19 Drupal8: Implement set UF locale/language (affects mailing tokens)

monishdeb opened the pull request:
Overview
----------------------------------------
How to reproduce:

1. Create a bilingual Drupal site, with the locale module, with URL-prefix language detection (example.org/fr/civicrm)
2. Enable multi-lingual CiviCRM, with two languages (ex: FR/EN)
3. Send a mailing in French, with the unsubscribe URL token


Expected result: the token should link to the page in French.

Before
----------------------------------------
The token isn't linking to the page in French

After
----------------------------------------
The token is linking to page in French

mlutfy commented:
I'm having trouble testing this. It's not crashing, but I don't see localized URLs. I think it's in part because of: https://lab.civicrm.org/dev/drupal/issues/21 ?

Not directly related: I think that Flexmailer is not implementing the call to `setLocale`, so I tested with flexmailer disabled, but that means that Mosaico does not support this yet.

monishdeb commented:
@mlutfy I have fixed the issue and for that need to do the following:
1. Overriden ```languageNegotiationURL()``` to do Drupal8 stuff to return language prefixed core path.
2. Fixed CRM_Utils_System_Drupal8::url() to accept base url
```php
$base = $absolute ? $config->userFrameworkBaseURL : $config->useFrameworkRelativeBase;
```
but need to do a hackish fix as currently ```\Drupal\Core\Url::fromUri()``` doesn't accept currently enabled language prefix, as a parameter, so that it, would later get appended to baseUrl path. Searched with other Drupal8 utility fn but got no luck. Please check my comment [here](https://github.com/civicrm/civicrm-core/pull/12152/files#diff-7b0caad195353c8c5d49bbf5f053daf6R310)

3. Overriden ```setUFLocale()```

eileenmcnaughton commented:
@mlutfy this is looking for re-review from you I think

eileenmcnaughton commented:
@mlutfy can you look at this again?

JoeMurray commented:
@mlutfy could you re-review?

JoeMurray commented:
@mlutfy could we trade you for a review of something you are working on? Eg https://github.com/civicrm/civicrm-core/pull/11085 ? Even https://github.com/civicrm/civicrm-core/pull/12527 which is 4.6?

mlutfy commented:
Hi Joe, sorry for the lack of response. Re-testing now.

mlutfy commented:
I'm running into a weird bug when a language does not have a language-prefix set.

For example, our users typically only use a language prefix for English:

- for French, do not use a language-prefix in URLs (`example.org/contact`)
- for English, use `/en` (`example.org/en/contact`).

When testing this on the CiviCRM backend dashboard, I see:

![capture d ecran de 2018-09-17 14-41-56](https://user-images.githubusercontent.com/254741/45643096-360de900-ba88-11e8-8d99-81b9820f5a0f.png)

i.e. the l10n menu fetches:  `https://civicrm/ajax/l10n-js/fr_CA?r=7cLZw`, because browsers automatically convert `//example` to `[current protocol]://example`.

Debugging the code, the `//` seems to come from `$config->useFrameworkRelativeBase`, which is set by the new implementation of `languageNegotiationURL()` ?

This seems to fix it for me:

```
        //url prefix
        if ($urlType == \Drupal\language\Plugin\LanguageNegotiation\LanguageNegotiationUrl::CONFIG_PATH_PREFIX) {
          if (!empty($language)) {
            if ($addLanguagePart && !empty($config['prefixes'][$language])) {  // <----- added !empty(..)
              $url .= $config['prefixes'][$language] . '/';
            }
            if ($removeLanguagePart) {
              $url = str_replace("/" . $config['prefixes'][$language] . "/", '/', $url);
            }
          }
        }
```

mlutfy commented:
I'm trying to compare the behaviour in D7 and D8, and would say that this patch definitely fixes a few things, and does not break anything (combined with the small fix in here:  https://github.com/civicrm/civicrm-core/pull/12152#issuecomment-422129533)

JoeMurray commented:
Monish could you incorporate improvements from @mlufty's comment and get jenkins to test? I think it is merge on pass at that point.

monishdeb commented:
@mlutfy thank you so much for you feedback and sharing your patch on multiple issues. I have rebased and updated the PR, and this is [commit](https://github.com/civicrm/civicrm-core/pull/12152/commits/6d27f36a8ccbc76e63397dc6eebd8308ee937ac1) for additional fixes. Did quick test on my D8 instance in local and it worked for me. Please have a look.

JoeMurray commented:
@eileenmcnaughton is this something that could be reviewed at the sprint (if it is still going on)?

eileenmcnaughton commented:
@JoeMurray I haven't found anyone with a d8 interest here :-( - @agileware might be able to assign someone? or @jackrabbithanna might do it or @dsnopek

JoeMurray commented:
Thanks, @eileenmcnaughton . We're willing to trade a review.

eileenmcnaughton commented:
@wannesderoy has #12635 open - so that might be a review trade opportunity @JoeMurray

dsnopek commented:
I'm quite confused on the desired behavior. So, if you're currently using the site is French at the time when you click "Send" on the mailing, then the mailing should include links to pages in French?

If so, that seems like kind of odd behavior, so I'm assuming I don't have that quite right...

The patch does however appear to be determining the current language solely from the URL? Since there's other ways for Drupal to determine the current language (in fact, it's pluggable, so contrib can add other ways too), it would seem better to ask Drupal to determine the language than just looking at the URL?

I think maybe I'm just misunderstanding this whole change in its entirety :-)

monishdeb commented:
@dsnopek actually it is more to do with what url type is configured in D8 whether domain or language prefix url need to be constructed on Civi enabled language. It is not deriving language from url but fetching as per D8 how url need to be constructed with locale or domain part using Civi current language ```$language = $this->getCurrentLanguage();```

mlutfy commented:
From the user's point of view, it's about making sure that tokens and URLs generated by CiviMail are in the language that the admin selected. i.e. if it's a mailing in French, then the unsubsribe email/token should point to the page in French.  If there are tokens with localized content, then the content should be in French.

mlutfy commented on CRM/Utils/System/Drupal8.php:
```diff
@@ -302,17 +302,23 @@ public function url(
+          'query' => $url['query'],
+          'fragment' => $fragment,
+        ])->toString()
+      );
```
Very oddly, this bit of code seems to be causing the Advance Search results to fail: run Search, then go to the second page of results. It shows a warning that the session has expired.

eileenmcnaughton commented:
test this please

eileenmcnaughton commented:
@mlutfy does that fix address what you hit?

eileenmcnaughton commented:
@mlutfy any chance you can take another look at this?

monishdeb commented:
ping @mlutfy @dsnopek

dsnopek commented:
I don't know that I understand this enough to review. All I can say is that looking at the code, hardcoding specific language negotiation plugins seems not quite right. If you have a site that adds some plugins, or maybe modifies one (it's possible to sub-class and replace a plugin) then weird things could happen.

JoeMurray commented:
So my sense is that the patch improves and fixes various things with the current implementation in Civi for i18n/l10n for a specific use case. Rewriting language negotiation in Civi for d8 along the lines suggested by @dsnopek seems like a good idea at some point but outside the scope of this fix to existing implementation. At this point, if @mlutfy can confirm that @monishdeb has addressed the most recent of the issues raised then I would encourage merging. This PR has already drifted to encompass solutions for various additional items and it would be good to merge it and create issues for any other new topics/issues found.

mlutfy commented:
I've been upgrading CiviCRM/Drupal and doing a fair amount of testing, and I think this PR fixes a fair amount of issues.

I have also been doing testing on single-language / English sites. I found two issue after upgrading to CiviCRM 5.10 / Drupal 8.6.10:

- If the site has a single language (English), but under Admin > Regional > Languages, the language negotiation was enabled and set to "Prefix", then some CiviCRM URLs will fail to load, because CiviCRM will try loading `/en/civicrm/[...].js` (ex: for the menu). However, that's rather obscure, because in our case, we have a bilingual site template and we return to single-language for monolingual clients.
- I had some new obscure errors accessing the CiviCRM REST API endpoint because if `if (!class_exists('Drupal'))`, which I fixed by patching to:

```
    if (!class_exists('Drupal') || !\Drupal::hasContainer()) {
      return NULL;
    }
```

Otherwise I would get:

```
"PHP message: PHP Fatal error:  Uncaught Drupal\Core\DependencyInjection\ContainerNotInitializedException: \Drupal::$container is not initialized yet.
\Drupal::setContainer() must be called with a real container. in [...]/core/lib/Drupal.php:130
```

@monishdeb Mind updating your PR to use the new condition?

Otherwise, we have been running this in production for the past few weeks and almost forgot it's there (we have 7 sites using it in production, a few more in dev).

MegaphoneJon commented:
@monishdeb I tested this as a potential fix for Drupal#38.

At first I got the same "Container not initialized" error @mlutfy mentions above.  After applying his suggested patch, my error became:

```
  [RuntimeException]                        
  Cannot resolve path using "cms.root.url" 
```

I think my patch will have to be applied over the top of yours.

monishdeb commented:
@MegaphoneJon that would be great. I'll cherry-pick your commit in my PR and will let you know when its ready to be tested?

eileenmcnaughton commented:
@monishdeb is this just pending on you adding the extra commit?

monishdeb commented:
@eileenmcnaughton @MegaphoneJon sorry for the delay. I have cherry-picked the commit from #13613 and updated this PR 

Please have a look. Thanks!

eileenmcnaughton commented:
@mlutfy my take on this is that it is potentially mergeable based on your comments & @MegaphoneJon ?

mlutfy commented:
I don't know if it's my esoteric setup causing issues, but I get : `The URI '/en/civicrm/ajax/dashboard' is invalid. You must use a valid URI scheme.`

When inspecting, `$base` = `/en/`.

It seems that `fromUri` expects "http://, https://, or an internal scheme eg: internal:/".

Until now, I was still using from an older patch:

```
      $url = str_replace('//dummy-link',
        "{$base}{$url['path']}",
        \Drupal\Core\Url::fromUri('//dummy-link', [
          'query' => $url['query'],
          'fragment' => $fragment,
        ])->toString()
      );
```

monishdeb commented:
@mlutfy that exposes an old issue where the language prefix in url path esp to Resource/AJAX urls makes it useless in D8-civi instance. A possible fix would be to identify and avoid such urls to have a language prefix. Now back to this PR, if I revert the fix from #13163 then it will cause an issue in CLI but including the commit will make internal urls to fail. Seems like we need to revert the last commit and then merge this PR as it addresses dev/drupal#19 ? I'll then submit a new PR for Drupal#38

mlutfy commented:
@monishdeb Sorry! I may have incorrectly applied the patch. I will do more testing first.

eileenmcnaughton commented:
@mlutfy any update?

eileenmcnaughton commented:
@mlutfy ping - trying to keep this one moving

mlutfy commented:
@monishdeb I did more testing. I still couldn't get a good result with the patch. URLs would render without the prefix URL.

You mention:

> "that exposes an old issue where the language prefix in url path esp to Resource/AJAX urls makes it useless in D8-civi instance. A possible fix would be to identify and avoid such urls to have a language prefix."

I'm not sure I understand? The menu needs the prefix in the URL, but it affects pretty much all links.

However, this works for me:

```
    $base = $absolute ? $config->userFrameworkBaseURL : 'internal:/';
```

i.e. `internal` instead of `base`, and the `/` is to prefix CiviCRM URLs, because d8 expects internal paths to start with a slash.

The [docs](https://api.drupal.org/api/drupal/core!lib!Drupal!Core!Url.php/function/Url%3A%3AfromUri/8.7.x) say:

"The internal: scheme should be avoided except when processing actual user input that may or may not correspond to a Drupal route. Normally use Url::fromRoute() for code linking to any any Drupal page."

... I'm not familiar enough with d8 to understand what "processing actual user input that may or may not correspond to a Drupal route" means.

dsnopek commented:
> ... I'm not familiar enough with d8 to understand what "processing actual user input that may or may not correspond to a Drupal route" means.

Oo, I don't understand most of this issue but I can contribute on this point. :-)

The Drupal 8 API wants you to use route names for everything (rather than internal paths, like in Drupal 7 and earlier). But a Drupal user has no idea about route names, so they'll always enter a path, for example, on a Link field. In that case, the Link field will look at the input and say, "oh, that's a relative path, so let's prepend 'internal:/' before we store that in the database". The alternative way to implement that would be to translate from the path into a route and store the route in the database, but that breaks user assumptions about how links work if they later change the page that lives at that path (for example, switching from a node to a view would have a different route name and parameters).

I don't know that that clarifies anything, though... Since this is all happening in code, you could conceivably use a route names, but I don't understand the changes here enough to suggest how or where you'd do that.

mlutfy commented:
Thanks @dsnopek, that definitely helps to understand!

If I understand correctly, using `internal:` this way is not how it was intended to be used, so it could have unwanted side-effects sooner or later.

Any idea why `base` and `internal` behave differently with the language prefix?

Here's a simple example, not CiviCRM-specific, that can be run using Drush on a site with two languages enabled:

```
print \Drupal\Core\Url::fromUri("base:/admin")->toString() . "\n";
print \Drupal\Core\Url::fromUri("internal:/admin")->toString() . "\n";
```

Output:

```
/admin
/fr/admin
```

mlutfy commented:
uhm, tried resolving the conflict using the github code browser, but that was probably not a good idea. :)

JoeMurray commented:
@monishdeb could you aim to do another review in next few days.

mlutfy commented:
There was another useful answer posted to SE, so I think we should go ahead and change the `fromUri` call, and merge this.

https://drupal.stackexchange.com/questions/278636/generating-urls-with-fromuri-and-the-language-prefix-base-vs-internal

eileenmcnaughton commented:
test this please

eileenmcnaughton commented:
@mlutfy are you saying you think a) this is mergeable as is but more changes will follow or b) this needs a change to be mergeable

mlutfy commented:
It needs a change (the `fromUri` call).

With that change, which I've been using it in production for a while, I would consider it mergeable. Of course, would be better if we had more feedback from other d8 folks, but having none, I would be tempted to create a new PR and get that merged.

eileenmcnaughton commented:
test this please

eileenmcnaughton commented:
@monishdeb can you make the minor change or @mlutfy can you merge & submit the change as a follow up & I'll merge it

eileenmcnaughton commented:
@monishdeb @mlutfy needs a rebase - maybe do a replacement pr @mlutfy unless @monishdeb gets there first - just keen to get this finished

monishdeb commented on CRM/Utils/System/Drupal8.php:
```diff
@@ -300,19 +300,20 @@ public function url(
     // Not all links that CiviCRM generates are Drupal routes, so we use the weaker ::fromUri method.
     try {
-      $url = \Drupal\Core\Url::fromUri("base:{$url['path']}", [
+      $url = \Drupal\Core\Url::fromUri("{$base}{$url['path']}", [
```
@mlutfy so that should be  ```\Drupal\Core\Url::fromUri("internal:{$url['path']}"``` right ?

mlutfy replied:
@monishdeb 

```
    $base = $absolute ? $config->userFrameworkBaseURL : 'internal:/';

    // Not all links that CiviCRM generates are Drupal routes, so we use the weaker ::fromUri method.
    try {
      $url = \Drupal\Core\Url::fromUri("{$base}{$url['path']}", [
      [...]
```

Before:

* CiviCRM generates non-localized URLs: `/civicrm/ajax/l10n-js/en_US?cid=131`

After:

* CiviCRM generates localized URLs: /en/civicrm/ajax/l10n-js/en_US?cid=131

nb: the `internal:/` prefix is important, otherwise URLs will be blank (Drupal8 URLs must start with a leading slash, but CiviCRM historally strips out the slash for other CMSes)

monishdeb replied:
I see thanks @mlutfy Will update the PR shortly

mlutfy commented:
I did a bit more testing, code review, and looks all good!

Thank you Monish !

monishdeb commented:
Thanks a lot, @mlutfy for getting this PR merged :)

eileenmcnaughton commented:
yay
